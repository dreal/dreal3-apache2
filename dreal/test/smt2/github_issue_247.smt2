(set-logic QF_NRA)


;--------------- SAMPLE CONTEXT FOR DESCRIBING THE BUG

(declare-fun a () Real)
(declare-fun b () Real)
(declare-fun c () Real)
(declare-fun d () Real)
(declare-fun e () Real)
(declare-fun f () Real)
(declare-fun Y () Real)


(assert (= a 8.768685282571274))
(assert (= b 29.761904761904763))
(assert (= c 5))
(assert (= d 15))
(assert (= e 0))
(assert (= f 10))
(assert (= Y 10))


(declare-fun Joker () Real)
(assert (= Joker 12.8313))


(define-fun absolute ( (val Real) ) Real
    (ite
        (>= val 0.0)
        val
        (- 0.0 val)
    )
)



(define-fun fun1 ( (x Real) ) Bool
    (<= (absolute (+ 10 (* -10 (- x a)))) Y)
)


;-------------  THE BUG/ISSUE

(declare-fun test_print1 () Bool)  ; test variable to print values
(assert
    (= test_print1
        (and (>= (+ a Joker) a) (<= (+ a Joker) b))
    )
)

;--- test_print1 EVALUATES TO TRUE


(declare-fun test_print2 () Bool)  ; test variable to print values
(assert
    (= test_print2
        (fun1 (+ a Joker))
    )
)

;--- test_print2 EVALUATES TO FALSE


(declare-fun test_print3 () Bool)  ; test variable to print values
(assert
    (= test_print3
        (and
            (and (>= (+ a Joker) a) (<= (+ a Joker) b)) ; equivalent to test_print1
            (fun1 (+ a Joker)) ; equivalent to test_print2
        )
    )
)

;--- test_print3, which is equivalent to the conjuction of test_print1 and test_print2 SHOULD EVALUATE TO FALSE, BUT IT EVALUATES TO TRUE !!!


(check-sat)
(get-model)
