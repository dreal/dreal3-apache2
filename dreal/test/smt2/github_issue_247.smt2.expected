delta-sat with delta = 0.001
(model
  (define-fun a () Real [8.7686852825712727, 8.7686852825712744])
  (define-fun b () Real [29.761904761904763, 29.761904761904766])
  (define-fun c () Real 5)
  (define-fun d () Real 15)
  (define-fun e () Real 0)
  (define-fun f () Real 10)
  (define-fun Y () Real 10)
  (define-fun Joker () Real [12.831299999999999, 12.831300000000001])
  (define-fun test_print1 () Bool true)
  (define-fun test_print2 () Bool false)
  (define-fun test_print3 () Bool false)
)
